### PR TITLE
Add middlewares for permission checking for v2 API

### PR DESCRIPTION
**Pull request by reasonerjt**
When the registry shifts from token auth to basic auth, we'll use the middleware to check permission.
This commit add middlewares for populate the artifact info and check
permission based on request to /v2/* api via security context

The change in `/src/core/xxx` can be disregarded because in the end all middlewares will be moved to `/src/server/middleware`

### CODE REVIEW

File: src/server/middleware/artifactinfo/artifact_info.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+var (
+	urlPatterns = []*regexp.Regexp{
+		middleware.V2ManifestURLRe, middleware.V2TagListURLRe, middleware.V2BlobUploadURLRe, middleware.V2BlobURLRe,
+	}
+)
+
+// Middleware gets the information of artifact via url of the request and inject it into the context
+func Middleware() func(http.Handler) http.Handler {
```
Comment (ywk253100) [inline src/server/middleware/artifactinfo/artifact_info.go]: This middleware can be merged into the manifest info middleware https://github.com/goharbor/harbor/blob/master/src/server/middleware/manifestinfo/manifest_info.go?

File: src/server/middleware/artifactinfo/artifact_info.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+
+			path := req.URL.Path
+			log.Debugf("in artifact info middleware, path: %s", path)
+			m, ok := parse(path)
+			if !ok {
+				next.ServeHTTP(rw, req)
```
Comment (ywk253100) [inline src/server/middleware/artifactinfo/artifact_info.go]: return?

File: src/server/middleware/artifactinfo/artifact_info_test.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/goharbor/harbor/src/core/middlewares/util"
+	"github.com/goharbor/harbor/src/server/middleware"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+)
+
+func TestMain(m *testing.M) {
```
Comment (ywk253100) [inline src/server/middleware/artifactinfo/artifact_info_test.go]: Why we need the `TestMain` here?

File: src/server/middleware/v2authz/authz.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/core/filter"
+	"github.com/goharbor/harbor/src/core/promgr"
+	ierror "github.com/goharbor/harbor/src/internal/error"
+	"github.com/goharbor/harbor/src/server/middleware"
+	reg_err "github.com/goharbor/harbor/src/server/registry/error"
+	"net/http"
+)
+
```
Comment (ywk253100) [inline src/server/middleware/v2authz/authz.go]: We should handle the request of blob mount

File: src/server/middleware/artifactinfo/artifact_info.go (hunk 1 above)
Comment (reasonerjt) [reply src/server/middleware/artifactinfo/artifact_info.go]: Yes,  but they are a little different, I opened  #10574 and will sync with Yan to merge them.

File: src/server/middleware/v2authz/authz.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+	if a, ok := middleware.ArtifactInfoFromContext(req.Context()); ok {
+		action := getAction(req)
+		if action == "" {
+			return nil
+		}
+		log.Debugf("action: %s, repository: %s", action, a.Repository)
+		p, err := rc.pm.Get(a.ProjectName)
```
Comment (heww) [inline src/server/middleware/v2authz/authz.go]: project may not found here

Comment (reasonerjt) [reply src/server/middleware/v2authz/authz.go]: If it's not found it will return an error

Comment (reasonerjt) [reply src/server/middleware/v2authz/authz.go]: Thanks I was thinking the pm will return an error if project is not found.

File: src/server/middleware/v2authz/authz.go (hunk 4 above)
Comment (reasonerjt) [reply src/server/middleware/v2authz/authz.go]: Yep, thanks.

File: src/server/middleware/artifactinfo/artifact_info_test.go (hunk 3 above)
Comment (reasonerjt) [reply src/server/middleware/artifactinfo/artifact_info_test.go]: We don't.  I'll remove.

**Comment by heww** [comment]
Is there another PR to apply these middlewares to v2?

**Comment by reasonerjt** [comment]
Yes, there will be.

**Pull request merged**